controlplane/kubeadm/internal/workload_cluster_etcd_test.go

```diff
@@ -0,0 +1,557 @@
+limitations under the License.
+*/
+
+package internal
+
+import (
+	"context"
+	"errors"
+	"testing"
+
+	. "github.com/onsi/gomega"
+
+	"go.etcd.io/etcd/clientv3"
+	pb "go.etcd.io/etcd/etcdserver/etcdserverpb"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/controlplane/kubeadm/internal/etcd"
+	fake2 "sigs.k8s.io/cluster-api/controlplane/kubeadm/internal/etcd/fake"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+func TestWorkload_EtcdIsHealthy(t *testing.T) {
+	g := NewWithT(t)
+
+	workload := &Workload{
+		Client: &fakeClient{
+			get: map[string]interface{}{
+				"kube-system/etcd-test-1": etcdPod("etcd-test-1", withReadyOption),
+				"kube-system/etcd-test-2": etcdPod("etcd-test-2", withReadyOption),
+				"kube-system/etcd-test-3": etcdPod("etcd-test-3", withReadyOption),
+				"kube-system/etcd-test-4": etcdPod("etcd-test-4"),
+			},
+			list: &corev1.NodeList{
+				Items: []corev1.Node{
+					nodeNamed("test-1", withProviderID("my-provider-id-1")),
+					nodeNamed("test-2", withProviderID("my-provider-id-2")),
+					nodeNamed("test-3", withProviderID("my-provider-id-3")),
+					nodeNamed("test-4", withProviderID("my-provider-id-4")),
+				},
+			},
+		},
+		etcdClientGenerator: &fakeEtcdClientGenerator{
+			client: &etcd.Client{
+				EtcdClient: &fake2.FakeEtcdClient{
+					EtcdEndpoints: []string{},
+					MemberListResponse: &clientv3.MemberListResponse{
+						Members: []*pb.Member{
+							{Name: "test-1", ID: uint64(1)},
+							{Name: "test-2", ID: uint64(2)},
+							{Name: "test-3", ID: uint64(3)},
+						},
+					},
+					AlarmResponse: &clientv3.AlarmResponse{
+						Alarms: []*pb.AlarmMember{},
+					},
+				},
+			},
+		},
+	}
+	ctx := context.Background()
+	health, err := workload.EtcdIsHealthy(ctx)
+	g.Expect(err).NotTo(HaveOccurred())
+
+	for _, err := range health {
+		g.Expect(err).NotTo(HaveOccurred())
+	}
+}
+
+func TestUpdateEtcdVersionInKubeadmConfigMap(t *testing.T) {
+	kubeadmConfig := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      kubeadmConfigKey,
+			Namespace: metav1.NamespaceSystem,
+		},
+		Data: map[string]string{
+			clusterConfigurationKey: `
+apiVersion: kubeadm.k8s.io/v1beta2
+kind: ClusterConfiguration
+etcd:
+  local:
+    dataDir: /var/lib/etcd
+    imageRepository: "gcr.io/k8s/etcd"
+    imageTag: "0.10.9"
+`,
+		},
+	}
+
+	g := NewWithT(t)
+	scheme := runtime.NewScheme()
+	g.Expect(corev1.AddToScheme(scheme)).To(Succeed())
+
+	tests := []struct {
+		name                  string
+		objs                  []runtime.Object
+		imageRepo             string
+		imageTag              string
+		expectErr             bool
+		expectedClusterConfig string
+	}{
+		{
+			name:      "returns error if unable to find kubeadm-config",
+			expectErr: true,
```

Comment:
This reads strange, like it's hard to tell why this test should fail. I think it's because `objs` is not being set, but can that be explicit?

Reply:
Yes. that's right. I can explicitly add `objs: []runtime.Object{}` or something like that.